jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/parser/WindowRateControl.java

```diff
@@ -0,0 +1,61 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2019 Mort Bay Consulting Pty. Ltd.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.http2.parser;
+
+import java.time.Duration;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/**
+ * <p>An implementation of {@link RateControl} that limits the number of
+ * events within a time period.</p>
+ * <p>Events are kept in a queue and for each event the queue is first
+ * drained of the old events outside the time window, and then the new
+ * event is added to the queue. If the size of the queue exceeds the max
+ * number of events then {@link #onEvent(Object)} returns {@code false}.</p>
+ */
+public class WindowRateControl implements RateControl
+{
+    private final Queue<Long> events = new ConcurrentLinkedQueue<>();
+    private final int maxEvents;
+    private final long window;
+
+    public WindowRateControl(int maxEvents, Duration window)
+    {
+        this.maxEvents = maxEvents;
+        this.window = window.toNanos();
+    }
+
+    @Override
+    public boolean onEvent(Object event)
+    {
+        long now = System.nanoTime();
+        while (true)
+        {
+            Long time = events.peek();
+            if (time == null)
+                break;
+            if (now - time < window)
+                break;
+            events.poll();
```

Comment:
This code is a race.  If two threads both peek the same event at 51, they will both do a poll at 56, removing 2 events instead of 1.

Reply:
Fixed by using `remove(Object)` - I think??? @sbordet @joakime